File: app/src/main/java/com/mainstreetcode/teammate/model/Game.java
/*
 * MIT License
 *
 * Copyright (c) 2019 Adetunji Dahunsi
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package com.mainstreetcode.teammate.model;

import androidx.room.Ignore;
import android.os.Parcel;
import androidx.annotation.NonNull;
import android.text.TextUtils;

import com.google.gson.JsonDeserializationContext;
import com.google.gson.JsonDeserializer;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonParseException;
import com.google.gson.JsonSerializationContext;
import com.google.gson.JsonSerializer;
import com.mainstreetcode.teammate.R;
import com.mainstreetcode.teammate.model.enums.Sport;
import com.mainstreetcode.teammate.persistence.entity.GameEntity;
import com.mainstreetcode.teammate.util.IdCache;
import com.mainstreetcode.teammate.util.ModelUtils;
import com.tunjid.androidbootstrap.recyclerview.diff.Differentiable;

import java.lang.reflect.Type;
import java.util.Arrays;
import java.util.Date;
import java.util.List;

import static com.mainstreetcode.teammate.util.ModelUtils.EMPTY_STRING;
import static com.mainstreetcode.teammate.util.ModelUtils.areNotEmpty;

/**
 * Event events
 */

public class Game extends GameEntity
        implements
        TeamHost,
        Model<Game>,
        HeaderedModel<Game>,
        ListableModel<Game> {

    @Ignore private static final IdCache holder = IdCache.cache(5);

    public Game(@NonNull String id, String name, String refPath, String score, String matchUp,
                String homeEntityId, String awayEntityId, String winnerEntityId,
                Date created, Sport sport, User referee, Team host, Event event, Tournament tournament,
                Competitor home, Competitor away, Competitor winner,
                int seed, int leg, int round, int homeScore, int awayScore,
                boolean ended, boolean canDraw) {
        super(id, name, refPath, score, matchUp, homeEntityId, awayEntityId, winnerEntityId, created, sport, referee, host, event, tournament, home, away, winner, seed, leg, round, homeScore, awayScore, ended, canDraw);
    }

    public static Game empty(Team team) {
        Sport sport = team.getSport();
        return new Game("", "", sport.refType(), "TBD", "", "", "", "",
                new Date(), sport, User.empty(), team, Event.empty(), Tournament.empty(),
                Competitor.empty(), Competitor.empty(), Competitor.empty(), 0, 0, 0, 0, 0, false, true);
    }

    public static Game withId(String id) {
        Game empty = empty(Team.empty());
        empty.setId(id);
        return empty;
    }

    protected Game(Parcel in) {
        super(in);
    }

    @Override
    public List<Item<Game>> asItems() {
        return Arrays.asList(
                Item.Companion.text(holder.get(0), 0, Item.NUMBER, R.string.game_competitors, this::getName, ignored -> {}, this),
                Item.Companion.number(holder.get(1), 1, Item.INPUT, R.string.game_home_score, () -> String.valueOf(getHomeScore()), this::setHomeScore, this),
                Item.Companion.number(holder.get(2), 2, Item.INPUT, R.string.game_away_score, () -> String.valueOf(getAwayScore()), this::setAwayScore, this),
                Item.Companion.number(holder.get(3), 3, Item.NUMBER, R.string.game_round, () -> String.valueOf(getRound()), ignored -> {}, this),
                Item.Companion.number(holder.get(4), 4, Item.NUMBER, R.string.game_leg, () -> String.valueOf(getLeg()), ignored -> {}, this)
        );
    }

    @Override
    public Item<Game> getHeaderItem() {
        return Item.Companion.text(EMPTY_STRING, 0, Item.IMAGE, R.string.team_logo, () -> "", url -> {}, this);
    }

    @Override
    public boolean areContentsTheSame(Differentiable other) {
        if (!(other instanceof Game)) return getId().equals(other.getId());
        Game casted = (Game) other;
        return getScore().equals(casted.getScore())
                && getHome().areContentsTheSame(casted.getHome())
                && getAway().areContentsTheSame(casted.getAway());
    }

    @Override
    public boolean hasMajorFields() {
        return areNotEmpty(getId(), getRefPath(), getScore()) && getHome().hasMajorFields() && getAway().hasMajorFields();
    }

    @Override
    public Object getChangePayload(Differentiable other) {
        return other;
    }

    @Override
    public boolean isEmpty() {
        return TextUtils.isEmpty(getId());
    }

    @Override
    public Team getTeam() { return getHost(); }

    @Override
    @SuppressWarnings("unchecked")
    public void update(Game updatedGame) {
        this.setId(updatedGame.getId());
        this.setScore(updatedGame.getScore());
        this.setMatchUp(updatedGame.getMatchUp());
        this.setCreated(updatedGame.getCreated());
        this.setLeg(updatedGame.getLeg());
        this.setSeed(updatedGame.getSeed());
        this.setRound(updatedGame.getRound());
        this.setHomeScore(updatedGame.getHomeScore());
        this.setAwayScore(updatedGame.getAwayScore());
        this.setEnded(updatedGame.isEnded());
        this.setCanDraw(updatedGame.getCanDraw());
        this.getSport().update(updatedGame.getSport());
        this.setHomeEntityId(updatedGame.getHomeEntityId());
        this.setAwayEntityId(updatedGame.getAwayEntityId());
        this.setWinnerEntityId(updatedGame.getWinnerEntityId());
        if (updatedGame.getReferee().hasMajorFields())
            this.getReferee().update(updatedGame.getReferee());
        if (updatedGame.getHost().hasMajorFields())
            this.getHost().update(updatedGame.getHost());
        if (updatedGame.getTournament().hasMajorFields())
            this.getTournament().update(updatedGame.getTournament());
        if (updatedGame.getHome().hasMajorFields() && this.getHome().hasSameType(updatedGame.getHome()))
            this.getHome().update(updatedGame.getHome());
        else this.setHome(updatedGame.getHome());
        if (updatedGame.getAway().hasMajorFields() && this.getAway().hasSameType(updatedGame.getAway()))
            this.getAway().update(updatedGame.getAway());
        else this.setAway(updatedGame.getAway());
        if (updatedGame.getWinner().hasMajorFields() && this.getWinner().hasSameType(updatedGame.getWinner()))
            this.getWinner().update(updatedGame.getWinner());
        else this.setWinner(updatedGame.getWinner());
        if (updatedGame.getEvent().hasMajorFields()) this.getEvent().update(updatedGame.getEvent());
    }

    @Override
    public int compareTo(@NonNull Game o) {
        int createdComparison = getCreated().compareTo(o.getCreated());

        return createdComparison != 0 ? createdComparison : getId().compareTo(o.getId());
    }

    @Override
    public int describeContents() {
        return 0;
    }

    public static final Creator<Game> CREATOR = new Creator<Game>() {
        @Override
        public Game createFromParcel(Parcel in) {
            return new Game(in);
        }

        @Override
        public Game[] newArray(int size) {
            return new Game[size];
        }
    };

    public static class GsonAdapter
            implements
            JsonSerializer<Game>,
            JsonDeserializer<Game> {

        private static final String ID_KEY = "_id";
        private static final String NAME = "name";
        private static final String REF_PATH = "refPath";
        private static final String SCORE = "score";
        private static final String MATCH_UP = "matchUp";
        private static final String CREATED_KEY = "created";
        private static final String SPORT_KEY = "sport";
        private static final String REFEREE = "referee";
        private static final String EVENT = "event";
        private static final String TOURNAMENT = "tournament";
        private static final String HOST = "host";
        private static final String HOME_ENTITY_ID = "homeEntity";
        private static final String AWAY_ENTITY_ID = "awayEntity";
        private static final String WINNER_ENTITY_ID = "winnerEntity";
        private static final String HOME = "home";
        private static final String AWAY = "away";
        private static final String WINNER = "winner";
        private static final String LEG = "leg";
        private static final String SEED = "seed";
        private static final String ROUND = "round";
        private static final String HOME_SCORE = "homeScore";
        private static final String AWAY_SCORE = "awayScore";
        private static final String ENDED = "ended";
        private static final String CAN_DRAW = "canDraw";

        @Override
        public JsonElement serialize(Game src, Type typeOfSrc, JsonSerializationContext context) {
            JsonObject body = new JsonObject();
            body.addProperty(ENDED, src.isEnded());
            body.addProperty(REF_PATH, src.getRefPath());
            body.addProperty(HOME_SCORE, src.getHomeScore());
            body.addProperty(AWAY_SCORE, src.getAwayScore());
            body.addProperty(HOME, src.getHome().getEntity().getId());
            body.addProperty(AWAY, src.getAway().getEntity().getId());
            body.addProperty(REFEREE, src.getReferee().isEmpty() ? null : src.getReferee().getId());
            return body;
        }

        @Override
        public Game deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context) throws JsonParseException {
            if (json.isJsonPrimitive()) {
                return new Game(json.getAsString(), "", "", "TBD", "", "", "", "",
                        new Date(), Sport.Companion.empty(), User.empty(), Team.empty(), Event.empty(), Tournament.empty(Team.empty()),
                        Competitor.empty(), Competitor.empty(), Competitor.empty(),
                        0, 0, 0, 0, 0, false, false);
            }

            JsonObject body = json.getAsJsonObject();

            String id = ModelUtils.asString(ID_KEY, body);
            String name = ModelUtils.asString(NAME, body);
            String refPath = ModelUtils.asString(REF_PATH, body);
            String score = ModelUtils.asString(SCORE, body);
            String matchUp = ModelUtils.asString(MATCH_UP, body);
            String homeEntityId = ModelUtils.asString(HOME_ENTITY_ID, body);
            String awayEntityId = ModelUtils.asString(AWAY_ENTITY_ID, body);
            String winnerEntityId = ModelUtils.asString(WINNER_ENTITY_ID, body);
            String created = ModelUtils.asString(CREATED_KEY, body);
            String sportCode = ModelUtils.asString(SPORT_KEY, body);

            int seed = (int) ModelUtils.asFloat(SEED, body);
            int leg = (int) ModelUtils.asFloat(LEG, body);
            int round = (int) ModelUtils.asFloat(ROUND, body);
            int homeScore = (int) ModelUtils.asFloat(HOME_SCORE, body);
            int awayScore = (int) ModelUtils.asFloat(AWAY_SCORE, body);
            boolean ended = ModelUtils.asBoolean(ENDED, body);
            boolean canDraw = ModelUtils.asBoolean(CAN_DRAW, body);

            Sport sport = Config.sportFromCode(sportCode);
            User referee = context.deserialize(body.get(REFEREE), User.class);
            Team host = context.deserialize(body.get(HOST), Team.class);
            Event event = context.deserialize(body.get(EVENT), Event.class);
            Tournament tournament = context.deserialize(body.get(TOURNAMENT), Tournament.class);
            Competitor home = context.deserialize(body.get(HOME), Competitor.class);
            Competitor away = context.deserialize(body.get(AWAY), Competitor.class);
            Competitor winner = body.has(WINNER) ? context.deserialize(body.get(WINNER), Competitor.class) : Competitor.empty();

            if (referee == null) referee = User.empty();
            if (host == null) host = Team.empty();
            if (event == null) event = Event.empty();
            if (home == null) home = Competitor.empty();
            if (away == null) away = Competitor.empty();
            if (tournament == null) tournament = Tournament.empty();

            return new Game(id, name, refPath, score, matchUp, homeEntityId, awayEntityId, winnerEntityId,
                    ModelUtils.parseDate(created), sport, referee, host, event, tournament,
                    home, away, winner, seed, leg, round, homeScore, awayScore, ended, canDraw);
        }
    }
}

File: app/src/main/java/com/mainstreetcode/teammate/model/Game.kt
/*
 * MIT License
 *
 * Copyright (c) 2019 Adetunji Dahunsi
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package com.mainstreetcode.teammate.model

import android.os.Parcel
import android.os.Parcelable
import android.text.TextUtils
import androidx.room.Ignore
import com.google.gson.JsonDeserializationContext
import com.google.gson.JsonDeserializer
import com.google.gson.JsonElement
import com.google.gson.JsonObject
import com.google.gson.JsonParseException
import com.google.gson.JsonSerializationContext
import com.google.gson.JsonSerializer
import com.mainstreetcode.teammate.R
import com.mainstreetcode.teammate.model.enums.Sport
import com.mainstreetcode.teammate.persistence.entity.GameEntity
import com.mainstreetcode.teammate.util.EMPTY_STRING
import com.mainstreetcode.teammate.util.IdCache
import com.mainstreetcode.teammate.util.areNotEmpty
import com.mainstreetcode.teammate.util.asBooleanOrFalse
import com.mainstreetcode.teammate.util.asFloatOrZero
import com.mainstreetcode.teammate.util.asStringOrEmpty
import com.mainstreetcode.teammate.util.noOp
import com.mainstreetcode.teammate.util.parseISO8601Date
import com.tunjid.androidbootstrap.recyclerview.diff.Differentiable
import java.lang.reflect.Type
import java.util.*

/**
 * Event events
 */

class Game : GameEntity,
        TeamHost,
        Model<Game>,
        HeaderedModel<Game>,
        ListableModel<Game> {

    constructor(
            id: String,
            name: String,
            refPath: String,
            score: String,
            matchUp: String,
            homeEntityId: String,
            awayEntityId: String,
            winnerEntityId: String,
            created: Date,
            sport: Sport,
            referee: User,
            host: Team,
            event: Event,
            tournament: Tournament,
            home: Competitor,
            away: Competitor,
            winner: Competitor,
            seed: Int,
            leg: Int,
            round: Int,
            homeScore: Int,
            awayScore: Int,
            ended: Boolean,
            canDraw: Boolean
    ) : super(id, name, refPath, score, matchUp, homeEntityId, awayEntityId, winnerEntityId, created, sport, referee, host, event, tournament, home, away, winner, seed, leg, round, homeScore, awayScore, ended, canDraw)

    private constructor(`in`: Parcel) : super(`in`)

    override val team: Team
        get() = host

    override val isEmpty: Boolean
        get() = TextUtils.isEmpty(id)

    override val headerItem: Item
        get() = Item.text(EMPTY_STRING, 0, Item.IMAGE, R.string.team_logo, { "" }, CharSequence::noOp)

    override fun asItems(): List<Item> = listOf(
            Item.text(holder[0], 0, Item.NUMBER, R.string.game_competitors, this::name, CharSequence::noOp),
            Item.number(holder[1], 1, Item.INPUT, R.string.game_home_score, homeScore::toString, this::setHomeScore),
            Item.number(holder[2], 2, Item.INPUT, R.string.game_away_score, awayScore::toString, this::setAwayScore),
            Item.number(holder[3], 3, Item.NUMBER, R.string.game_round, round::toString, CharSequence::noOp),
            Item.number(holder[4], 4, Item.NUMBER, R.string.game_leg, leg::toString, CharSequence::noOp)
    )

    override fun areContentsTheSame(other: Differentiable): Boolean =
            if (other !is Game) diffId == other.diffId else score == other.score
                    && home.areContentsTheSame(other.home)
                    && away.areContentsTheSame(other.away)

    override fun hasMajorFields(): Boolean =
            areNotEmpty(id, refPath, score) && home.hasMajorFields() && away.hasMajorFields()

    override fun getChangePayload(other: Differentiable): Any? = other

    override fun update(updated: Game) {
        this.id = updated.id
        this.score = updated.score
        this.matchUp = updated.matchUp
        this.created = updated.created
        this.leg = updated.leg
        this.seed = updated.seed
        this.round = updated.round
        this.homeScore = updated.homeScore
        this.awayScore = updated.awayScore
        this.isEnded = updated.isEnded
        this.canDraw = updated.canDraw
        this.sport.update(updated.sport)
        this.homeEntityId = updated.homeEntityId
        this.awayEntityId = updated.awayEntityId
        this.winnerEntityId = updated.winnerEntityId
        if (updated.referee.hasMajorFields())
            this.referee.update(updated.referee)
        if (updated.host.hasMajorFields())
            this.host.update(updated.host)
        if (updated.tournament.hasMajorFields())
            this.tournament.update(updated.tournament)
        if (updated.home.hasMajorFields() && this.home.hasSameType(updated.home))
            this.home.update(updated.home)
        else
            this.home = updated.home
        if (updated.away.hasMajorFields() && this.away.hasSameType(updated.away))
            this.away.update(updated.away)
        else
            this.away = updated.away
        if (updated.winner.hasMajorFields() && this.winner.hasSameType(updated.winner))
            this.winner.update(updated.winner)
        else
            this.winner = updated.winner
        if (updated.event.hasMajorFields()) this.event.update(updated.event)
    }

    override fun compareTo(other: Game): Int =
            compareValuesBy(this, other, Game::created, Game::id)

    override fun describeContents(): Int = 0

    class GsonAdapter : JsonSerializer<Game>, JsonDeserializer<Game> {

        override fun serialize(src: Game, typeOfSrc: Type, context: JsonSerializationContext): JsonElement {
            val body = JsonObject()
            body.addProperty(ENDED, src.isEnded)
            body.addProperty(REF_PATH, src.refPath)
            body.addProperty(HOME_SCORE, src.homeScore)
            body.addProperty(AWAY_SCORE, src.awayScore)
            body.addProperty(HOME, src.home.entity.id)
            body.addProperty(AWAY, src.away.entity.id)
            body.addProperty(REFEREE, if (src.referee.isEmpty) null else src.referee.id)
            return body
        }

        @Throws(JsonParseException::class)
        override fun deserialize(json: JsonElement, typeOfT: Type, context: JsonDeserializationContext): Game {
            if (json.isJsonPrimitive) {
                return Game(json.asString, "", "", "TBD", "", "", "", "",
                        Date(), Sport.empty(), User.empty(), Team.empty(), Event.empty(), Tournament.empty(Team.empty()),
                        Competitor.empty(), Competitor.empty(), Competitor.empty(),
                        0, 0, 0, 0, 0, ended = false, canDraw = false)
            }

            val body = json.asJsonObject

            val id = body.asStringOrEmpty(ID_KEY)
            val name = body.asStringOrEmpty(NAME)
            val refPath = body.asStringOrEmpty(REF_PATH)
            val score = body.asStringOrEmpty(SCORE)
            val matchUp = body.asStringOrEmpty(MATCH_UP)
            val homeEntityId = body.asStringOrEmpty(HOME_ENTITY_ID)
            val awayEntityId = body.asStringOrEmpty(AWAY_ENTITY_ID)
            val winnerEntityId = body.asStringOrEmpty(WINNER_ENTITY_ID)
            val created = body.asStringOrEmpty(CREATED_KEY)
            val sportCode = body.asStringOrEmpty(SPORT_KEY)

            val seed = body.asFloatOrZero(SEED).toInt()
            val leg = body.asFloatOrZero(LEG).toInt()
            val round = body.asFloatOrZero(ROUND).toInt()
            val homeScore = body.asFloatOrZero(HOME_SCORE).toInt()
            val awayScore = body.asFloatOrZero(AWAY_SCORE).toInt()
            val ended = body.asBooleanOrFalse(ENDED)
            val canDraw = body.asBooleanOrFalse(CAN_DRAW)

            val sport = Config.sportFromCode(sportCode)
            val referee: User = context.deserialize<User>(body.get(REFEREE), User::class.java)
                    ?: User.empty()
            val host: Team = context.deserialize<Team>(body.get(HOST), Team::class.java)
                    ?: Team.empty()
            val event: Event = context.deserialize<Event>(body.get(EVENT), Event::class.java)
                    ?: Event.empty()
            val tournament: Tournament = context.deserialize<Tournament>(body.get(TOURNAMENT), Tournament::class.java)
                    ?: Tournament.empty()
            val home: Competitor = context.deserialize<Competitor>(body.get(HOME), Competitor::class.java)
                    ?: Competitor.empty()
            val away: Competitor = context.deserialize<Competitor>(body.get(AWAY), Competitor::class.java)
                    ?: Competitor.empty()
            val winner =
                    if (body.has(WINNER)) context.deserialize(body.get(WINNER), Competitor::class.java)
                            ?: Competitor.empty()
                    else Competitor.empty()

            return Game(id, name, refPath, score, matchUp, homeEntityId, awayEntityId, winnerEntityId,
                    created.parseISO8601Date(), sport, referee, host, event, tournament,
                    home, away, winner, seed, leg, round, homeScore, awayScore, ended, canDraw)
        }

        companion object {

            private const val ID_KEY = "_id"
            private const val NAME = "name"
            private const val REF_PATH = "refPath"
            private const val SCORE = "score"
            private const val MATCH_UP = "matchUp"
            private const val CREATED_KEY = "created"
            private const val SPORT_KEY = "sport"
            private const val REFEREE = "referee"
            private const val EVENT = "event"
            private const val TOURNAMENT = "tournament"
            private const val HOST = "host"
            private const val HOME_ENTITY_ID = "homeEntity"
            private const val AWAY_ENTITY_ID = "awayEntity"
            private const val WINNER_ENTITY_ID = "winnerEntity"
            private const val HOME = "home"
            private const val AWAY = "away"
            private const val WINNER = "winner"
            private const val LEG = "leg"
            private const val SEED = "seed"
            private const val ROUND = "round"
            private const val HOME_SCORE = "homeScore"
            private const val AWAY_SCORE = "awayScore"
            private const val ENDED = "ended"
            private const val CAN_DRAW = "canDraw"
        }
    }

    companion object {

        @Ignore
        private val holder = IdCache(5)

        fun empty(team: Team): Game {
            val sport = team.sport
            return Game(
                    id = "",
                    name = "",
                    refPath = sport.refType(),
                    score = "TBD",
                    matchUp = "",
                    homeEntityId = "",
                    awayEntityId = "",
                    winnerEntityId = "",
                    created = Date(),
                    sport = sport,
                    referee = User.empty(),
                    host = team,
                    event = Event.empty(),
                    tournament = Tournament.empty(),
                    home = Competitor.empty(),
                    away = Competitor.empty(),
                    winner = Competitor.empty(),
                    seed = 0,
                    leg = 0,
                    round = 0,
                    homeScore = 0,
                    awayScore = 0,
                    ended = false,
                    canDraw = true
            )
        }

        fun withId(id: String): Game {
            val empty = empty(Team.empty())
            empty.id = id
            return empty
        }

        @JvmField
        val CREATOR: Parcelable.Creator<Game> = object : Parcelable.Creator<Game> {
            override fun createFromParcel(`in`: Parcel): Game = Game(`in`)

            override fun newArray(size: Int): Array<Game?> = arrayOfNulls(size)
        }
    }
}
